### PR TITLE
fix(room_preview): add fallback server names when possible

### PR DESCRIPTION
**Pull request by jmartinesp**
If no server names are provided for the room summary request and the room's server name doesn't match the current user's server name, add the room alias/id server name as a fallback value. This seems to fix room preview through federation, when there is a server name in either the room id or alias.

Also, when getting a summary for a room list item, if it's an invite one, add the server name of the inviter's user id as another possible fallback.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk/src/room_preview.rs [hunk 1]
```diff
@@ -372,3 +375,92 @@ async fn search_for_room_preview_in_room_directory(
...
 }
+
+// Make sure the server name of the room id/alias is
+// included in the list of server names to send if no server names are provided
+fn ensure_server_names_is_not_empty(
+    own_server_name: Option<&ServerName>,
+    server_names: Vec<OwnedServerName>,
+    room_or_alias_id: &RoomOrAliasId,
+) -> Vec<OwnedServerName> {
+    let Some(own_server_name) = own_server_name else { return server_names };
```
Comment (poljar) [inline crates/matrix-sdk/src/room_preview.rs:386]:
```suggestion
    let Some(own_server_name) = own_server_name else { return server_names };
    
```

File: crates/matrix-sdk/src/room_preview.rs [hunk 2]
```diff
@@ -372,3 +375,92 @@ async fn search_for_room_preview_in_room_directory(
...
+    if server_names.is_empty() {
+        if let Some(server_name) = room_or_alias_id.server_name() {
+            if server_name != own_server_name {
+                return server_names.into_iter().chain(vec![server_name.to_owned()]).collect();
+            }
+        }
+    }
```
Comment (poljar) [inline crates/matrix-sdk/src/room_preview.rs:393]:
```suggestion
    }
    
```

File: crates/matrix-sdk/src/room_preview.rs [hunk 3]
```diff
@@ -372,3 +375,92 @@ async fn search_for_room_preview_in_room_directory(
...
+fn ensure_server_names_is_not_empty(
+    own_server_name: Option<&ServerName>,
+    server_names: Vec<OwnedServerName>,
+    room_or_alias_id: &RoomOrAliasId,
+) -> Vec<OwnedServerName> {
+    let Some(own_server_name) = own_server_name else { return server_names };
+    if server_names.is_empty() {
+        if let Some(server_name) = room_or_alias_id.server_name() {
+            if server_name != own_server_name {
+                return server_names.into_iter().chain(vec![server_name.to_owned()]).collect();
```
Comment (poljar) [inline crates/matrix-sdk/src/room_preview.rs:390]: You could have just pushed the additional name to the vec and avoid the early return.

```suggestion
                server_names.push(server_name.to_owned());
```

**Review by poljar — APPROVED** [review]
Left some nits, looks good.

File: crates/matrix-sdk/src/room_preview.rs (hunk 3 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/room_preview.rs:390]: `server_names` is not mutable. I can make it so, but then I'd be creating a new `Vec` even when no changes are needed.

Comment (poljar) [reply crates/matrix-sdk/src/room_preview.rs:390]: But you wouldn't, you would return the same `Vec` that was passed to the method, some times you would mutate it and sometimes you won't.

The `collect()` creates a brand new `Vec`.

Comment (jmartinesp) [reply crates/matrix-sdk/src/room_preview.rs:390]: Oh, I completely misunderstood how this works then (you'd actually be creating a different type on Kotlin, i.e.). I'll change that, thanks!

**Pull request merged**